packages/core/test/tags.test.ts

```diff
@@ -0,0 +1,54 @@
+import { createMachine } from '../src';
+
+describe('tags', () => {
+  it('supports tagging states', () => {
+    const machine = createMachine({
+      initial: 'green',
+      states: {
+        green: {
+          tags: ['go']
+        },
+        yellow: {
+          tags: ['go'],
+          on: {
+            TIMER: 'red'
+          }
+        },
+        red: {
+          tags: ['stop']
+        }
+      }
+    });
+
+    expect(machine.initialState.tags.has('go')).toBeTruthy();
+    expect(machine.transition('yellow', 'TIMER').tags.has('go')).toBeFalsy();
+  });
+
+  it('supports tags in compound states', () => {
+    const machine = createMachine({
+      initial: 'red',
+      states: {
+        green: {
+          tags: ['go']
+        },
+        yellow: {},
+        red: {
+          tags: ['stop'],
+          initial: 'walk',
+          states: {
+            walk: {
+              tags: ['crosswalkLight']
+            },
+            wait: {
+              tags: ['crosswalkLight']
+            }
+          }
+        }
+      }
+    });
+
+    expect(machine.initialState.tags.has('go')).toBeFalsy();
```

Comment:
I'm wondering if we shouldn't provide a specialized `.has` on the state itself so we could support "typed tags" in the future. See my comment here: https://github.com/davidkpiano/xstate/discussions/2124#discussioncomment-644216 . The idea is pretty much the same as with typestates.

Reply:
I think that's a good idea; let's save it for a separate PR

Reply:
I agree that we don't have to introduce "typed tags" right from the start but if that's a direction that we are going to then we could already introduce the `has` method to `State` (it would basically just call `this.tags.has(tag)` and nothing more than that.

Why introducing it already? To keep the documented interface more future-proof - using the underlying set directly limits our possibilities and when we finally introduce `has` then people will have to refactor their existing code to switch to it.